Migrate Calculations to DX

toropok opened the pull request:
## Description of the issue/feature this PR addresses

This PR migrates the AT Calculation to Dexterity for Python 3 compatibility.

## Current behavior before PR

Calculation is a content type based on the Archetypes framework

## Desired behavior after PR is merged

Calculation is a content type based on the Dexterity framework

--
I confirm I have tested this PR thoroughly and coded it according to [PEP8][1]
and [Plone's Python styleguide][2] standards.

[1]: https://www.python.org/dev/peps/pep-0008
[2]: https://docs.plone.org/develop/styleguide/python.html

toropok committed:
init

toropok committed:
init

toropok committed:
init

toropok committed:
init

toropok committed:
init

toropok committed:
merge 2.x

toropok committed:
update

toropok committed:
Merge branch '2.x' into dx-calculations

toropok committed:
update

toropok committed:
linter warnings removed

toropok committed:
Merge branch '2.x' into dx-calculations

toropok committed:
linter warnings

toropok committed:
import fix

toropok committed:
updat

toropok committed:
Merge branch '2.x' into dx-calculations

toropok committed:
fix

toropok committed:
fix

toropok committed:
fix

toropok committed:
fix

toropok committed:
Merge branch '2.x' into dx-calculations

toropok committed:
fix

toropok committed:
fix

toropok committed:
merge 2.x

toropok committed:
merge 2.x

toropok committed:
fix

toropok committed:
merge 2.x

toropok committed:
Merge branch '2.x' into dx-calculations

toropok committed:
fix

toropok committed:
build css

toropok committed:
update

toropok committed:
update

toropok committed:
Merge branch '2.x' into dx-calculations

toropok committed:
linter fixes

toropok committed:
tests fix

toropok committed:
upgrade fix

toropok committed:
test & fixes

toropok commented:
Hi @ramonski @xispa,

Here's my preliminary delivery for the calc migration! 😊 This migration turned out to be quite a tough nut compared to the previous ones—no chance it would’ve worked without the invaluable help of @Bugerman58! 🚀

Currently, we’re stuck at a point where the tests are failing. Specifically:

We’re not using HistoryAwareReferenceField as expected.
Was there any thought about using an alternative?
Or is there no longer a need to maintain version references?
Looking forward to your input!

ramonski commented:
Hi Leonid, hi Anton,
many thanks for your continuous efforts and support that we really appreciate!
Yes, indeed, you picked a tough nut with the calculations 🤯
Regarding the `HistoryAwareReferenceWidget`, Jordi and I discussed once that we want to drop it and maybe lookup the values from the Snapshot storage? And even combine this with our `UIDReferenceField`?
Anyhow, we'll find a solution for this for sure.
Thanks again for your good work and we'll review as soon as possible!

xispa commented:
Excellent work @toropok and @Bugerman58 :1st_place_medal: . I can imagine how challenging the migration of Calculation has been!. Your amazing work is really appreciated! Willing to review the Pull Request already and maybe we open another one later for the removal of the `HistoryAware` relic!

toropok committed:
Merge branch '2.x' into dx-calculations

xispa commented:
Hi @toropok , tests are failing https://github.com/senaite/senaite.core/actions/runs/11948903328/job/33307294669

Can you please take a look before I start to review?

toropok commented:
> Hi @toropok , tests are failing https://github.com/senaite/senaite.core/actions/runs/11948903328/job/33307294669
> 
> Can you please take a look before I start to review?

Hi @xispa 

Sure, I'll fix the failed Python tests (they only test object validation) and get back to you. The remaining failing tests are doctests, which are related to HistoryAwareReferenceField.

toropok committed:
Merge branch '2.x' into dx-calculations

toropok committed:
merge 2.x

toropok committed:
merge 2.x

toropok committed:
fix

toropok committed:
Merge branch '2.x' into dx-calculations

toropok committed:
linter

toropok committed:
fix

toropok committed:
fix

toropok commented:
@xispa 

Sorry for the long pause.

- I’ve added `FormulaValidator`, and it works! 🙂
- Fixed the Python tests — all tests are now passing successfully.

That leaves us with two remaining issues that need your guidance and vision:

1. HistoryAware behavior for `UIDReferenceField`: All the failing doctests are related to this feature.
2. Formula wildcards: How are they supposed to work? It seems like they’ve never worked as intended.

xispa commented on src/senaite/core/api/analysisservice.py:
```diff
@@ -0,0 +1,198 @@
+
+        for calc_uid in calc_uids:
+            # Get the calculation object
+            calc = api.get_object_by_uid(calc_uid)
```
I get the following traceback on sample creation form after selecting a profile with tests that have calculations assigned:

```
Traceback (innermost last):
  Module ZPublisher.WSGIPublisher, line 176, in transaction_pubevents
  Module ZPublisher.WSGIPublisher, line 385, in publish_module
  Module ZPublisher.WSGIPublisher, line 288, in publish
  Module ZPublisher.mapply, line 85, in mapply
  Module ZPublisher.WSGIPublisher, line 63, in call_object
  Module bika.lims.decorators, line 51, in decorator
  Module bika.lims.browser.analysisrequest.add2, line 723, in __call__
  Module bika.lims.browser.analysisrequest.add2, line 1221, in ajax_recalculate_records
  Module bika.lims.browser.analysisrequest.add2, line 1413, in get_unmet_dependencies_info
  Module senaite.core.api.analysisservice, line 157, in get_service_dependencies_for
  Module senaite.core.api.analysisservice, line 85, in get_calculation_dependants_for
  Module senaite.core.api.analysisservice, line 62, in calc_dependants_gen
  Module bika.lims.api, line 858, in get_object_by_uid
  Module bika.lims.api, line 487, in fail
APIError: No object found for UID c15d28b73a314c6bad5ce8d4becd569e
```

Problem gets solved when legacy's `get_backreferences` (for AT types) is used at [L57](https://github.com/senaite/senaite.core/pull/2600/files#diff-891391bd7fece7c1e4b116246b12db904f624fb31b3be788d81c7d52d2a499cbR57) instead of `get_backrefs`:

```python
(Pdb++) service
<AnalysisService at /senaite/bika_setup/bika_analysisservices/analysisservice-8>
(Pdb++) get_backrefs(service, relationship="CalculationDependentServices")
['c15d28b73a314c6bad5ce8d4becd569e']
(Pdb++) get_backreferences(service, relationship="CalculationDependentServices")
['6fb9450c2bd445e0aa4b272c4aef4a96']
(Pdb++) api.get_object_by_uid("6fb9450c2bd445e0aa4b272c4aef4a96")
<Calculation at /senaite/setup/calculations/calculation-7>
```

However, note that `Calculation` content type is now a DX. Therefore, `get_backreferences` should no longer be used in favour of `get_backrefs`. We probably have a problem here when updating the references in the upgrade step. 

Further investigation is required.

xispa replied:
Yes, got the problem, see comment below at https://github.com/senaite/senaite.core/pull/2600/files#diff-df4a46baf178ed9a06d8c236e808a1893d8518c77290bd45e0e605820792fa81R2773

xispa commented on src/senaite/core/upgrade/v02_06_000.py:
```diff
@@ -2725,3 +2727,125 @@ def reindex_specs(tool):
+    target.setFormula(src.getFormula())
+    target.setTestParameters(src.getTestParameters() or [])
+    target.setTestResult(src.getTestResult() or "")
+    target.setDependentServices(src.getDependentServices() or [])
```
Note that `DependentServices` field has the attribute `relationship` assigned to `CalculationDependentServices` here: https://github.com/senaite/senaite.core/pull/2600/files#diff-ac00ca15357fd79f94dff708a40950c317c3e001a4ffda4ad1f0f6d8e07c7a80R320

This means that when you set the value for this field (an `UIDReferenceField`), the system will automatically assign a back reference to this target object (a `Calculation`) on each object assigned (a list of `AnalysisService` objects). In a nutshell: the system will store the `UID` of the `Calculation` object as an `Annotation` on each `AnalysisService` being assigned. However, at this point, the `UID` of the `Calculation` object is still temporary, hasn't been yet replaced  by the original one at https://github.com/senaite/senaite.core/pull/2600/files#diff-df4a46baf178ed9a06d8c236e808a1893d8518c77290bd45e0e605820792fa81R2783

So, you need to either:

(a) call `setDependentServices` after `migrator.copy_uid` 
(b) manually update back references for fields with `relationship` attribute declared after `migrator.copy_uid`.

I think (a) is easier, but might be confusing and forces the developer to know beforehand which fields need to be set after the UID replacement. The implementation of (b) is not as easy as (a), but in my opinion, more robust and easier to follow.

What do you think @ramonski ?

ramonski replied:
Yes, I would do it manually as well, e.g. like this (not tested!!):

```python
from senaite.core.schema.uidreferencefield import get_backref_storage
from bika.lims.browser.fields.uidreferencefield import get_backreferences

key = "CalculationDependentServices"
old_backrefs = get_backreferences(src, relationship=key)

new_storage = get_backref_storage(target)
new_storage[key] = PersistentList()
new_backrefs = new_storage[key]

for ref in old_backrefs:
    new_backrefs.append(api.get_uid(ref))
```

Phew, not sure if I missed here something. Best you inspect with a PDB further Leonid!

xispa commented on src/senaite/core/browser/form/adapters/calculation.py:
```diff
@@ -18,17 +18,179 @@
+FIELD_INTERIM_VALUE = "form.widgets.interims.{}.widgets.value"
+
+
+def deep_get(dictionary, *keys):
```
Is this strictly necessary?. The data `dict` is not that much nested and for what I see, you only use `deep_get` to extract the value of the formula from the form, that is something that can be easily done like this:

```python
form = data.get("form")
formula = form.get(FIELD_FORMULA) or ""
```

xispa commented on src/senaite/core/upgrade/v02_06_000.py:
```diff
@@ -2725,3 +2727,125 @@ def reindex_specs(tool):
+    # NOTE: always convert string values to unicode for dexterity fields!
+    target.title = api.safe_unicode(src.Title() or "")
+    target.description = api.safe_unicode(src.Description() or "")
+    target.setInterimFields(src.getInterimFields() or [])
```
With AT's `InterimFields` field, non-required subfields without data were omitted (e.g. `choices`, `result_type`, etc.). However, DX's `InterimFields` expect these values to be set even if non-required.

Basically, after running the upgrade step you get this:

![Captura de 2024-12-11 09-24-58](https://github.com/user-attachments/assets/e9988959-b63f-4caf-9902-3b9d9ce58c98)

Note the `<NO_VALUE>` and `Missing: <NO_VALUE>`. Reason is that the keys `choice` and `result_type` are missing. Before you set the interim fields to the new `Calculation` object you need fill the gaps with defaults first: `''` (empty) for `choices` and `numeric` for `result_type`.

xispa commented on src/senaite/core/browser/form/adapters/calculation.py:
```diff
@@ -18,17 +18,179 @@
+            None).validate(
+                {"formula": deep_get(data, "form", FIELD_FORMULA) or "",
+                 "interims":
+                 [{"keyword": i} for i in self.get_interim_keywords(data)]})
```
Really nice, so system "detects" if an interim field or service for a given keyword does exist each time a form value is edited.... definitely this helps a lot!!

However, would you mind to simplify the code a bit? something like below would be easier to read imho:

```python
form = data.get("form")

validator = FormulaValidator(self.context, self.request, self, ICalculationSchema, None)
value = {
    "formula": form.get(FIELD_FORMULA) or "",
    "interims": self.get_interim_fields(data)
}
errors = validator.validate(value)
```

Note I would rely on a `self.get_interim_fields(data)` that returns well-formed interim field records, containing not only with the keyword, but also the rest of subfields (you never know the subfields the validator will rely on).

xispa commented on src/senaite/core/browser/form/adapters/calculation.py:
```diff
@@ -18,17 +18,179 @@
+from senaite.core.content.calculation import ICalculationSchema
+from senaite.core.validators.formula import FormulaValidator
+
+interim_keyword_regex = re.compile(r"interims\.(\d+)\.widgets\.keyword$")
```
For consistency reasons, better declare these constants as follows:

```python
INTERIM_KEYWORD_RX = re.compile(r"interims\.(\d+)\.widgets\.keyword$")
```

xispa commented on src/senaite/core/browser/form/adapters/calculation.py:
```diff
@@ -18,17 +18,179 @@
+        self.add_update_field(FIELD_TEST_RESULT, result)
         return self.data
+
+    def process_keywords(self, data):
```
This function is really difficult to grasp. For what I see it does four things:

1) Extracts the keywords from the formula
2) Extracts the `UID`s for the keywords with counterpart services
3) Updates the `DependentServices` field with those extracted `UID`s
4) Returns a `dict` where the keys are the keywords and the value is the value set for the keyword in the "Test Parameters" section.

I suggest to remove this function and rely on functions that clearly state what they do instead, e.g. (not tested):

```python
INTERIM_KEYWORD_RX = re.compile(r"\[([^.^\]]+)\]")

def get_formula(self, data):
    """Returns the formula of the form
    """
    form = data.get("form")
    formula = form.get(FIELD_FORMULA) or ""
    # cleanup line breaks
    formula = " ".join(formula.split("\r\n"))
    return formula  

def get_formula_keywords(self, data):
    """Returns the keywords from the formula
    """
    formula = self.get_formula(data)
    # get the keywords without brackets
    keywords = INTERIM_KEYWORD_RX.findall(formula)
    # remove duplicates
    return list(set(keywords))

def get_formula_services(self, data):
    """Returns the keywords counterpart service UIDs from the formula
    """
    keywords = self.get_formula_keywords(data)
    query = {
        "portal_type": "AnalysisService",
        "getKeyword": keywords
    }
    brains = api.search(query, SETUP_CATALOG)
    return list(map(api.get_uid, brains))

def get_formula_interim_keywords(self, data):
    """Returns the keywords without service counterpart from the formula
    """
    keywords = self.get_formula_keywords(data)
    services = self.get_formula_services(data)
    services = [api.get_object(service) for service in services]
    service_keywords = [service.getKeyword() for keyword in services]
    return filter(lambda key: key not in service_keywords, keywords)
```

btw, note that with `re.compile(r"\[([^.^\]]+)\]")` you get the interim values without square brackets directly.

xispa commented on src/senaite/core/browser/static/js/senaite.core.calculation.edit.js:
```diff
@@ -0,0 +1,74 @@
+            return rawTestValue;
+        }
+    });
+});
```
@ramonski wouldn't be better to have this js as a component inside `webpack`?

If so, the `calculation_edit.pt` and change in `viewlets/configure.zcml` could then be removed.

ramonski replied:
This seems to me like a JS only loaded for the edit page right?
Then we could move the whole logic probably as an edit form adapter in `senaite.core.browser.form`.

toropok replied:
> This seems to me like a JS only loaded for the edit page right? Then we could move the whole logic probably as an edit form adapter in `senaite.core.browser.form`.

@ramonski 

This custom JavaScript code provides several UX benefits, including:

- Preventing users from adding `test_parameters` rows manually to the `DataGridWidget`;
- Listening to formula update events and dynamically reevaluating formulas;
- Making certain fields in the `TestParameters` DataGridWidget partially read-only mode;

I don't think we can achieve such behavior without it by implementing only form adapter.

xispa commented on src/senaite/core/schema/interimsfield.py:
```diff
@@ -0,0 +1,144 @@
+            u"label_interim_keyword",
+            default=u"Keyword"
+        ),
+        required=False,
```
Make it `required=True`

xispa commented on src/senaite/core/schema/interimsfield.py:
```diff
@@ -0,0 +1,144 @@
+            u"label_interim_title",
+            default=u"Field title"
+        ),
+        required=False,
```
Make it `required=True`

xispa commented on src/senaite/core/schema/interimsfield.py:
```diff
@@ -0,0 +1,144 @@
+        ),
+        source="senaite.core.vocabularies.resulttypes",
+        default="numeric",
+        required=False,
```
Better make it `required=True`. An option "No value" will be available for selection otherwise:

![Captura de 2024-12-11 11-27-46](https://github.com/user-attachments/assets/d88b4391-01c4-44d1-abd5-7181188f4116)

xispa commented on src/senaite/core/browser/viewlets/configure.zcml:
```diff
@@ -368,4 +368,13 @@
       permission="zope2.View"
       layer="senaite.core.interfaces.ISenaiteCore" />
 
+  <browser:viewlet
```
This might not be necessary if `calculation.edit.js` is added as a component inside `webpack`

ramonski replied:
Yes, I would try to completely remove the JS and provide a custom edit form adapter

xispa commented on src/senaite/core/browser/viewlets/templates/calculation_edit.pt:
```diff
@@ -0,0 +1 @@
+<script tal:attributes="src string:${view/site_url}/++plone++senaite.core.static/js/senaite.core.calculation.edit.js"></script>
```
This template might not be necessary if calculation.edit.js is added as a component inside webpack

xispa commented on src/senaite/core/content/calculation.py:
```diff
@@ -0,0 +1,602 @@
+        auto_append=True)
+    interims = InterimsField(
+        title=_(u"label_calculation_interims",
+                default=u"Calculation Interim Fields"),
```
Please use "Variables" for title and description, like we have for `AnalysisService`: https://github.com/senaite/senaite.core/blob/2.x/src/bika/lims/content/analysisservice.py#L112-L119

I think is cleaner. From a user perspective, "interim" sounds confusing imo.

xispa commented on src/senaite/core/content/calculation.py:
```diff
@@ -0,0 +1,602 @@
+                      u"Interim fields defined above, as well as any services "
+                      u"that this calculation depends on to "
+                      u"calculate results."),
+        value_type=DataGridRow(schema=ITestParameterRecord),
```
An styling and cosmetic issue, but I think that indenting the title and description fields like this makes the code easier to read:

```python
    test_parameters = DataGridField(
        title=_(
            u"title_calculation_test_parameters",
            default=u"Test Parameters"
        ),
        description=_(
            u"description_calculation_test_parameters",
            default=u"To test the calculation, enter values here for all "
                    u"calculation parameters. This includes Interim fields "
                    u"defined above, as well as any services that this "
                    u"calculation depends on to calculate results."
        ),
        value_type=DataGridRow(schema=ITestParameterRecord),
        required=False,
        default=[]
    )
```

xispa commented on src/senaite/core/workflow/calculation/guards.py:
```diff
@@ -0,0 +1,49 @@
+    brains = api.search(query, SETUP_CATALOG)
+    for brain in brains:
+        service = api.get_object(brain)
+        calc = service.getCalculation()
```
To not wake up the `Calculation` object unnecessarily, you can rely on `getRawCalculation` instead:

```python
calc_uid = service.getRawCalculation()
if calc_uid == calculation.UID():
    return False
```

xispa reviewed (changes requested):
Pheew, a lot of work here... @toropok you must be exhausted. I've done a first review and requested some changes though.

Haven't gone through the validators details yet, but I think you can simplify a lot there. I feel are overwhelming, will review them in the next round.

I've seen tests that are failing are because of the `HistoryAwareReferenceField`. Don't worry too much about this for now.

xispa commented:
> 1. HistoryAware behavior for `UIDReferenceField`: All the failing doctests are related to this feature.

Yes, I've seen. I've requested some changes, we will handle this `HistoryAware` behavior afterwards.

> 2. Formula wildcards: How are they supposed to work? It seems like they’ve never worked as intended.

They are supposed to work, yes. For instance, if the user types this formula:

```python
[CHOL] + [FAT] if [CHOL] >= [CHOL.LDL] else "Invalid"
```

where:

- `CHOL` is the keyword of a service (e.g. Cholesterol)
- `FAT` is the keyword of a saervice (e.g. Fat)
- `<service_keyword>.LDL` is the wildcard representing the value of the field `LowerDetectionLimit` for the prefixed service

Therefore, if the result of `CHOL` is below the `LowerDetectionLimit`, the result of the calculation will be `Invalid`.

See e.g. the doctests here: https://github.com/senaite/senaite.core/blob/2.x/src/senaite/core/tests/test_calculations.py#L83-L137

toropok committed:
Merge branch '2.x' into dx-calculations

toropok committed:
fix of generating proper dendendent services uids on migration

toropok committed:
fix deactivation guard

toropok committed:
move js to webpack

toropok committed:
migration iinterims fix & naming issue

toropok committed:
fixes

toropok committed:
fix minor adapter revision

toropok committed:
fix dotted params substitution

toropok committed:
Merge branch '2.x' into dx-calculations

toropok commented:
> They are supposed to work, yes. For instance, if the user types this formula:

Yeah, I see now. But it's never worked before for calculating test results when wildcards are used, at least.

ramonski commented:
Hi Leo,
can you grant access for me on this branch, so that I can support you with direct code contribution?

toropok commented:
> Hi Leo, can you grant access for me on this branch, so that I can support you with direct code contribution?

Hi @ramonski 

Yes, sure! But how to do so? :)

ramonski commented:
I see you added me already as a contributor, thanks! Let's check if this works out

ramonski committed:
Merge branch '2.x' into dx-calculations

ramonski committed:
Merge branch '2.x' into dx-calculations

ramonski committed:
Changelog updated

ramonski committed:
Removed bika.lims.calculation.edit.js

ramonski committed:
Production JS

ramonski committed:
Ensure correct order in test parameters

ramonski committed:
initialize test parameters

ramonski committed:
Imports ordered

ramonski committed:
Fix unicode decode errors in validation chain

ramonski committed:
Minor changes

ramonski committed:
compatibility for additional wildcards of #2682

ramonski committed:
re-added default imports

ramonski committed:
do not remove spaces in formula

ramonski committed:
convert formula to unicode

ramonski committed:
minor formatting

ramonski committed:
Remove repository tool profile

ramonski committed:
Change versioning API to return the snapshot version

ramonski committed:
Generated JS files added

ramonski committed:
upgrade step version

xispa committed:
Merge branch '2.x' into dx-calculations

ramonski committed:
Merged branch 2.x

ramonski committed:
Production JS built

ramonski committed:
Moved upgrade step to 2.7

ramonski committed:
Merge branch '2.x' into dx-calculations

ramonski committed:
Rebuild Production JS

ramonski committed:
Merge branch '2.x' into dx-calculations

ramonski committed:
Production JS

ramonski committed:
Updated upgrade step

ramonski committed:
Handle version wrapper objects in API

ramonski committed:
Do not allow to call restricted methods on a version wrapper

ramonski committed:
additional attributes

ramonski committed:
Remove Calculation HistoryAwareReferenceField

ramonski committed:
Handle versioned references

ramonski committed:
Removed unused code

ramonski committed:
Fixed test

ramonski committed:
Fixed test

ramonski committed:
minor change

ramonski committed:
Use api.get_tool instead of getToolByName

ramonski committed:
pass Formula into api.create method

ramonski committed:
Fixed test

ramonski committed:
Fixed test

ramonski committed:
Use aq_base function instead of attribute

ramonski committed:
convert falsy interim values to empty string

ramonski committed:
fixed test

ramonski committed:
Fixed test

ramonski committed:
Fixed test

ramonski committed:
Fixed test

ramonski committed:
Fixed tests

ramonski committed:
Fixed test

ramonski committed:
Fixed tests

ramonski committed:
Fixed test

ramonski committed:
removed unused import

ramonski committed:
Fixed test

ramonski committed:
Check for none or empty string instead

ramonski committed:
removed guard transition caching

ramonski committed:
Add missing interim `value` with empty string instead of 0

ramonski committed:
Re-add caching

ramonski committed:
fixed test

ramonski committed:
Fixed test

ramonski committed:
Fixed test

ramonski commented:
Hi @toropok, @Bugerman58,

the tests are finally passing:) so I guess we're almost there!

However, some crucial part is still missing:
In the migration step, we need to fetch all the stored versions from `portal_repository` and create snapshots according their versions to make the version wrapper work accordingly (in case we have an older instance before we introduced the snapshots).

Also, we need to clean up all of the stored versions of https://github.com/plone/Products.CMFEditions and eventually the references created.

Also, I get currently this error for some of the existing calculations:
![SCR-20250527-lsvx](https://github.com/user-attachments/assets/d82b6c18-e366-459e-83b3-5f529bbe84fc)

Maybe you can have a look where this error might come from and I take care of the rest.

ramonski committed:
removed _

toropok commented:
Hi, @ramonski !

You did an enormous job—fantastic work!

We’ll definitely take care of the upgrade step soon.

Regarding the error you encountered in the screenshot: from what I can see, it occurred during the 'eval' stage of the calculation (`calculation.py:156`)

```python
def calculate_formula(formula="", parameters={}, imports=None):
    result = "Failure"

    try:
        formula = f_formatter.format(formula, parameters)
        result = eval(formula, getGlobals(imports))
```
—

an unhandled exception was thrown. Could you please copy and share the text of the exception? It should be in the Test Result field in the screenshot.

ramonski commented:
Hi Leonid,
Thanks for your prompt response.
Unfortunately, I see nothing in the logs and the Test field only shows `Unspecified exception: invalid syntax`. But let me put a PDB on that code line you pointed out

ramonski commented:
Unfortunately, the error is not related to the formula evaluation. In fact this just works just fine:
<img width="1520" alt="SCR-20250527-olhw" src="https://github.com/user-attachments/assets/a6863425-ed38-40e9-a8a9-cfc5f019ecc2" />

It seems to be related to one of the datagrid fields, but does neither show in the logs nor in the UI any hints.

Another thing I have seen is that I was somehow able to remove the entire data grid for the "Additional Python Libraries", so that it is no longer possible to enter any values (even after reload it stays empty):

<img width="1708" alt="SCR-20250527-omjh" src="https://github.com/user-attachments/assets/1f89a6eb-0f32-48de-b840-2f277b70921c" />

ramonski commented:
So in fact it seems to be related to the interim DGF somehow:

```python
> /Users/rbartl/develop/buildout/eggs/z3c.form-3.7.1-py2.7.egg/z3c/form/field.py(314)

 291         def extract(self):
 292             """See interfaces.IWidgets"""
 293             data = {}
 294             errors = ()
 295             for name, widget in self.items():
 296                 if widget.mode == interfaces.DISPLAY_MODE:
 297                     continue
 298                 value = widget.field.missing_value
 299                 try:
 300                     widget.setErrors = self.setErrors
 301                     raw = widget.extract()
 302                     if raw is not interfaces.NO_VALUE:
 303                         value = interfaces.IDataConverter(widget).toFieldValue(raw)
 304                     widget.ignoreRequiredOnValidation = self.ignoreRequiredOnExtract
 305                     zope.component.getMultiAdapter(
 306                         (self.content,
 307                          self.request,
 308                          self.form,
 309                          getattr(widget, 'field', None),
 310                          widget),
 311                         interfaces.IValidator).validate(value)
 312                 except (zope.interface.Invalid,
 313                         ValueError, MultipleErrors) as error:
 314  ->                 import pdb; pdb.set_trace()
 315                     view = zope.component.getMultiAdapter(
 316                         (error, self.request, widget, widget.field,
 317                          self.form, self.content), interfaces.IErrorViewSnippet)
 318                     view.update()
 319                     if self.setErrors:
 320                         widget.error = view
 321                     errors += (view,)
 322                 else:
 323                     name = widget.__name__
 324                     data[name] = value
 325             for error in self.validate(data):
 326                 view = zope.component.getMultiAdapter(
 327                     (error, self.request, None, None, self.form, self.content),
 328                     interfaces.IErrorViewSnippet)
 329                 view.update()
 330                 errors += (view,)
 331             if self.setErrors:
 332                 self.errors = errors
 333             return data, errors
(Pdb++) errors
(<ErrorViewSnippet for RequiredMissing>,)
(Pdb++) snippet = errors[0]
(Pdb++) snippet.
content        createMessage  field          message        request        widget
context        error          form           render         update
(Pdb++) snippet.render()
u'\n  <div class="error">Required input is missing.</div>\n\n'
(Pdb++) widget
<TextWidget 'form.widgets.interim_fields.AA.widgets.title'>
(Pdb++)
```

ramonski committed:
Remove underscore from name

toropok commented:
> So in fact it seems to be related to the interim DGF somehow:
> 
> ```python
> (Pdb++) snippet.render()
> u'\n  <div class="error">Required input is missing.</div>\n\n'
> (Pdb++) widget
> <TextWidget 'form.widgets.interim_fields.AA.widgets.title'>
> (Pdb++)
> ```

Thanks! That gave me a hint: the AutoAppend row has no title value, but it's marked as required.
Okay, I'll check it soon (and Imports as well)

toropok committed:
fix empty import DGF

toropok committed:
Syntax error added to eval exceptions list

toropok commented:
HI @ramonski 

I cant reproduce an error from your first screenshot. Can you give any instructions for steps to get there?
 
> Maybe you can have a look where this error might come from and I take care of the rest.

ramonski commented:
> HI @ramonski
> 
> I cant reproduce an error from your first screenshot. Can you give any instructions for steps to get there?
> 
> > Maybe you can have a look where this error might come from and I take care of the rest.

Hi Leonid,
Thanks for your support and efforts!
I actually noticed that behavior for some migrated calculations, but not all (which makes this bug quite dubious).
The only thing I found out is the PDB output from above. Unfortunately, there are no other hints I've seen so far.
However, I try to find out more as soon as I return to this migration.

toropok commented:
hmmm.. got it! Thanks!! may be thats somehow caused by migrations. We will see that soon!

toropok committed:
Merge branch '2.x' into dx-calculations

ramonski commented:
Hi Leonid, I'm going to test this PR now once again in an existing DB.

toropok commented:
> Hi Leonid, I'm going to test this PR now once again in an existing DB.

Hi Ramon!

You're so welcome :) we were trying to make migration work :)

ramonski commented:
> > Hi Leonid, I'm going to test this PR now once again in an existing DB.
> 
> Hi Ramon!
> 
> You're so welcome :) we were trying to make migration work :)

Ah, I see. Unfortunately, the migraion is failing at the moment:
```
Traceback (innermost last):
  Module ZPublisher.WSGIPublisher, line 176, in transaction_pubevents
  Module ZPublisher.WSGIPublisher, line 385, in publish_module
  Module ZPublisher.WSGIPublisher, line 288, in publish
  Module ZPublisher.mapply, line 85, in mapply
  Module ZPublisher.WSGIPublisher, line 63, in call_object
  Module Products.GenericSetup.tool, line 1135, in manage_doUpgrades
  Module Products.GenericSetup.upgrade, line 185, in doStep
  Module senaite.core.upgrade, line 39, in wrap_func_args
  Module senaite.core.upgrade.v02_07_000, line 128, in migrate_calculations_to_dx
  Module Products.GenericSetup.tool, line 375, in runImportStepFromProfile
  Module Products.GenericSetup.tool, line 1323, in _doRunImportStep
   - __traceback_info__: typeinfo
  Module Products.CMFCore.exportimport.typeinfo, line 222, in importTypesTool
  Module Products.GenericSetup.utils, line 933, in importObjects
   - __traceback_info__: portal_types
  Module Products.GenericSetup.utils, line 929, in importObjects
   - __traceback_info__: types/Calculation
  Module Products.GenericSetup.utils, line 530, in _importBody
  Module Products.CMFCore.exportimport.typeinfo, line 61, in _importNode
  Module Products.GenericSetup.utils, line 762, in _initProperties
ValueError: undefined property 'add_permission'
```

I'll check once again when you updated the codebase.

Also remember to rebuild the webpack sources (just to be sure after the last 2.x merge).

toropok committed:
move remove_at_portal_types to utils

toropok committed:
lint fix

toropok committed:
lint fix

toropok commented:
@ramonski 

Ramon,

migration routine fixed

ramonski commented on src/senaite/core/tests/doctests/ARAnalysesField.rst:
```diff
@@ -49,11 +49,11 @@ Functional Helpers:
-    ...     if api.is_at_content(content):
-    ...         content.processForm()
-    ...     elif api.is_dexterity_content(content):
-    ...         modified(content)
```
I guess this one needs to go back

toropok replied:
my bad, missed on merge. now will set it back

ramonski committed:
Rebuild Production JS

ramonski commented:
Just encountered this after I copied an existing sample with a changed calculation analysis inside:

```
> /Users/rbartl/develop/buildout/src/senaite.core/src/bika/lims/browser/analyses/view.py(1077)

1067         def get_interim_choices(self, interim):
1068             """Parse the interim choices field
1069             """
1070
1071             import pdb; pdb.set_trace()
1072             choices = interim.get("choices")
1073             if not choices:
1074                 return None
1075             items = choices.split("|")
1076             pairs = map(lambda item: item.strip().split(":"), items)
1077  ->         return OrderedDict(pairs)
ValueError: need more than 1 value to unpack
(Pdb++) choinces
*** NameError: name 'choinces' is not defined
(Pdb++) interim
{u'apply_wide': 'False', u'keyword': 'x', u'title': 'X Factor', u'value': '1', u'choices': 'None', u'allow_empty': 'False', u'report': 'False', u'hidden': 'False', u'result_type': 'numeric', u'unit': 'None'}
(Pdb++) choices
'None'
```

Also note the `unit`, which is also set to `'None'`.

Maybe we must set the default to an empty string instead?
Or handle these values correctly in the calculation `getInterimFields`?
https://github.com/toropok/senaite.core/blob/c998927feac7f71d22b3f32649dcd7bbdd8e65ef/src/bika/lims/browser/analyses/view.py#L1110

toropok committed:
revert test version

toropok committed:
Merge branch 'dx-calculations' of https://github.com/toropok/senaite.core into dx-calculations

toropok commented:
Aha I see 

I'll fix that tonight!

toropok commented:
@ramonski 

hmmm, seems that we already assign empty string as a default values to `choices` and `unit` fields. I'm trying to figure out where from this conversion to 'None' string comes from.

ramonski commented:
> @ramonski
> 
> hmmm, seems that we already assign empty string as a default values to `choices` and `unit` fields. I'm trying to figure out where from this conversion to 'None' string comes from.

Probably this comes from the migration. I'm going to check that and report back.
Furthermore, we have to keep in mind that we have to create snapshots for all the stored versions in `portal_repository` in the upgrade step. This is required to make the versionwrapper work consistently with the old versions (in case we're upgrading an instance before snapshots were in place)

toropok committed:
fix new InterimField setter: replace None with '' value

ramonski committed:
Create new snapshots from versioned objects

toropok committed:
fix new InterimField setter: replace None with '' value

ramonski commented:
That's strange.
I migrated an existing DB successfully with the provided upgrade step.
However, as soon as update one of the new DX Calculations and add interim fields, I get this error again:

<img width="552" alt="SCR-20250608-shis" src="https://github.com/user-attachments/assets/146b0063-f52e-4a57-a293-7c11dd8d0ffd" />


And then, the others are also no longer working.

ramonski committed:
Fix test

ramonski committed:
Change comment

ramonski commented:
Hi Leonid,
the issue mentioned in https://github.com/senaite/senaite.core/pull/2600#issuecomment-2912744546 is coming from the [edit form adapter](https://github.com/toropok/senaite.core/blob/d3753d0c1587c9c84e628ba6c8a1646566bd1bd8/src/senaite/core/browser/form/adapters/calculation.py)
I'm investigating at the moment, where this might happen

toropok commented:
> Hi Leonid, the issue mentioned in [#2600 (comment)](https://github.com/senaite/senaite.core/pull/2600#issuecomment-2912744546) is coming from the [edit form adapter](https://github.com/toropok/senaite.core/blob/d3753d0c1587c9c84e628ba6c8a1646566bd1bd8/src/senaite/core/browser/form/adapters/calculation.py) I'm investigating at the moment, where this might happen

Hi Ramon!

We tried several times to reproduce this error in our test environment and to get the same result you encountered, but without success — everything works fine after the migration. Maybe you can give us more details to model your setup locally.

However, we did find that the `getInterimFields()` method of the `Analysis` object returns boolean fields as strings (e.g., 'False'), and at some point, the keys in the this dictionary become Unicode strings. We've been trying to locate this tricky spot for the whole day without success. 

When calling `calc.getInterimFields()` from the debug console or in a Calculation view context, it returns the correct object. However, when the getters are invoked within the Analysis context, things are get messy.

`Calculation.getInterimFields()` (or Plone debug console):
```python
[{'apply_wide': False, 'keyword': u'TC5', 'title': u'TC5', 'value': '', 'choices': '', 'allow_empty': False, 'report': False, 'hidden': False, 'result_type': 'numeric', 'unit': ''}]
```
`Analysis.getInterimFields()`  view it becomes:
```python
[{u'apply_wide': u'False', u'keyword': u'TC5', u'title': u'TC5', u'value': u'', u'choices': u'', u'allow_empty': u'False', u'report': u'False', u'hidden': u'False', u'result_type': u'numeric', u'unit': u''}]
```

magical thing

ramonski commented:
<img width="880" alt="SCR-20250609-kxab" src="https://github.com/user-attachments/assets/bccf6c07-fc65-4421-9f5a-15ebd1878983" />

Ok, this seems like one of the issues.
Another one might be related to https://github.com/senaite/senaite.core/pull/2737, as with ReactJS 19 the widgets can no longer be expected to be loaded on `DOMContentLoaded` event...

toropok commented:
Thanks! now I got it. We'll fix it

toropok committed:
fix multivalue dependant service field assigment

ramonski committed:
Migrate dependent services

ramonski committed:
Handle ReactJS Components value setting

ramonski committed:
Set dependent services

ramonski committed:
Production JS

ramonski committed:
Production JS

ramonski committed:
Fix raw test input wrapper

ramonski commented:
Hi Leonid,
Ok, after some brain twists I think I got it working now (see my last commits).

I also added a comment in https://github.com/senaite/senaite.core/pull/2600/commits/5adcd743837b850e658eb22353cc77a80528665d, so maybe we find a better solution for this, e.g. the `input` or custom `modified` event maybe?

@xispa: Can you give it a try as well?

And is it a correct handling if we have, e.g. a calculation in a sample with version `v6` and the newest version of the calulation is `v7`, that we get `v7` if we retract the analysis?

toropok commented:
hi Ramon

we've this value passing from the front now, did we update something wrong?

<img width="1438" alt="image" src="https://github.com/user-attachments/assets/60a9d09d-0fed-4baf-8ecf-7b6c961954b8" />
<img width="1438" alt="image" src="https://github.com/user-attachments/assets/f416f346-40ef-4eaa-919b-c468868e8fb7" />

ramonski commented:
> hi Ramon
> 
> we've this value passing from the front now, did we update something wrong?
> 
> <img alt="image" width="1438" src="https://private-user-images.githubusercontent.com/14867014/453261511-60a9d09d-0fed-4baf-8ecf-7b6c961954b8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDk1MzEwODksIm5iZiI6MTc0OTUzMDc4OSwicGF0aCI6Ii8xNDg2NzAxNC80NTMyNjE1MTEtNjBhOWQwOWQtMGZlZC00YmFmLThlY2YtN2I2Yzk2MTk1NGI4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA2MTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNjEwVDA0NDYyOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPThlZWEwZDljZjFmZDYxYjNlODc0OWI5ZWZiMTM4NmRkMDkzNmNkMTE0ZGQwYmU5M2QzNWE5OGQ0NTY2ZWUzZjImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.n10XY6w0xAzfeY-12uPkv7_FfticTsMU5rQEJGyH5Gs"> <img alt="image" width="1438" src="https://private-user-images.githubusercontent.com/14867014/453262129-f416f346-40ef-4eaa-919b-c468868e8fb7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDk1MzEwODksIm5iZiI6MTc0OTUzMDc4OSwicGF0aCI6Ii8xNDg2NzAxNC80NTMyNjIxMjktZjQxNmYzNDYtNDBlZi00ZWFhLTkxOWItYzQ2ODg2OGU4ZmI3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA2MTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNjEwVDA0NDYyOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRkN2QzMjhlOTA1ZGYyOTA4ZGU5ZWUzMjhlMDk4MjVkYmZkZDA3MjM1Njg1OTYyOGEyZDg0ZjFhNjdiMDQ1NGUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.lqvdQ95Kp_7DhhAi4xEWHKopD5X9r1ohUWFpsWA4DXA">

Ok, got it. I'll take care of it.

ramonski committed:
Comment only

ramonski committed:
Provide value for a textarea istead of a reference

xispa commented:
> Hi Leonid, Ok, after some brain twists I think I got it working now (see my last commits).
> 
> I also added a comment in [5adcd74](https://github.com/senaite/senaite.core/commit/5adcd743837b850e658eb22353cc77a80528665d), so maybe we find a better solution for this, e.g. the `input` or custom `modified` event maybe?
> 
> @xispa: Can you give it a try as well?
> 
> And is it a correct handling if we have, e.g. a calculation in a sample with version `v6` and the newest version of the calulation is `v7`, that we get `v7` if we retract the analysis?

Sure, will test as well within this week

ramonski committed:
Added fallback for reference field value

ramonski committed:
Production JS

ramonski commented:
Hi Leonid,
Thanks for pointing this out. I fixed the issue with https://github.com/senaite/senaite.core/pull/2600/commits/d49e6e9f216313cc031fd70bdbd12afd9e6faa7d.
The reason for this issue was, that this field behaves in "hidden" mode like a normal `textarea` field and not like a reference field, where this this conversion is already done  here: https://github.com/toropok/senaite.core/blob/c60acc903154a5abb8a44b5c2fe2a6bf2520895a/webpack/app/components/editform.js#L603

However, I added now a fallback as well to handle the raw reference value as well.

ramonski committed:
Merge branch '2.x' into dx-calculations

ramonski commented:
Hi Leonid,
we need to convert the interim values properly, e.g. `'False'` should be empty string while `'True'` should be probably `'on'`.
Otherwise they are not handled in the sample analysis listing and appear hidden:

```python
[25] > /Users/rbartl/develop/buildout/src/senaite.core/src/bika/lims/browser/analyses/view.py(1122)_folder_item_calculation()
-> interim_unit = interim_field.get("unit", "")
> /Users/rbartl/develop/buildout/src/senaite.core/src/bika/lims/browser/analyses/view.py(1122)

1077         def _folder_item_calculation(self, analysis_brain, item):
1078             """Set the analysis' calculation and interims to the item passed in.
1079
1080             :param analysis_brain: Brain that represents an analysis
1081             :param item: analysis' dictionary counterpart that represents a row
1082             """
1083
1084             if not self.has_permission(ViewResults, analysis_brain):
1085                 # Hide interims and calculation if user cannot view results
1086                 return
1087
1088             is_editable = self.is_analysis_edition_allowed(analysis_brain)
1089
1090             # calculation
1091             calculation = self.get_calculation(analysis_brain)
1092             calculation_uid = api.get_uid(calculation) if calculation else ""
1093             calculation_title = api.get_title(calculation) if calculation else ""
1094             calculation_link = get_link_for(calculation) if calculation else ""
1095             item["calculation"] = calculation_uid
1096             item["Calculation"] = calculation_title
1097             item["replace"]["Calculation"] = calculation_link or _("Manual")
1098
1099             if is_editable and calculation:
1100                 url = analysis_brain.getURL()
1101                 item["after"]["Result"] = item["after"].get("Result") or ""
1102                 item["after"]["Result"] += get_link(
1103                     "{}/action/recalculate".format(url),
1104                     value="<i class='small text-secondary fas fa-sync'></i>",
1105                     title=t(_("Recalculate")), css_class="listing-ajax-action")
1106
1107             # Set interim fields. Note we add the key 'formatted_value' to the list
1108             # of interims the analysis has already assigned.
1109             analysis_obj = self.get_object(analysis_brain)
1110             interim_fields = analysis_obj.getInterimFields() or list()
1111             import pdb; pdb.set_trace()
1112
1113             # Copy to prevent to avoid persistent changes
1114             interim_fields = deepcopy(interim_fields)
1115             for interim_field in interim_fields:
1116                 interim_keyword = interim_field.get("keyword", "")
1117                 if not interim_keyword:
1118                     continue
1119
1120                 interim_value = interim_field.get("value", "")
1121                 interim_allow_empty = interim_field.get("allow_empty") == "on"
1122  ->             interim_unit = interim_field.get("unit", "")
1123
1124                 # Get the interim's formatted value
1125                 interim_formatted = self.get_formatted_interim(interim_field)
1126                 interim_field["formatted_value"] = interim_formatted
1127
1128                 # Update the item with the interim
1129                 item[interim_keyword] = interim_field
1130                 item["class"][interim_keyword] = "interim"
1131
1132                 # render the unit after the interim field
1133                 if interim_unit:
1134                     formatted_interim_unit = format_supsub(interim_unit)
1135                     item["after"][interim_keyword] = self.render_unit(
1136                         formatted_interim_unit)
1137
1138                 # Note: As soon as we have a separate content type for field
1139                 #       analysis, we can solely rely on the field permission
1140                 #       "senaite.core: Field: Edit Analysis Result"
1141                 if is_editable:
1142                     if self.has_permission(
1143                             FieldEditAnalysisResult, analysis_brain):
1144                         item["allow_edit"].append(interim_keyword)
1145
1146                 # Add this analysis' interim fields to the interim_columns list
1147                 interim_hidden = interim_field.get("hidden", False)
1148                 if not interim_hidden:
1149                     interim_title = interim_field.get("title")
1150                     self.interim_columns[interim_keyword] = interim_title
1151
1152                 # Does interim's results list needs to be rendered?
1153                 choices = self.get_interim_choices(interim_field)
1154                 if choices:
1155                     multi = self.is_multi_interim(interim_field)
1156
1157                     # Ensure empty option is available if no default value is set
1158                     if not interim_value and not multi:
1159                         # allow empty selection and flush default value
1160                         interim_allow_empty = True
1161
1162                     # Generate the display list
1163                     # [{"ResultValue": value, "ResultText": text},]
1164     ...
(Pdb++) interim_field
{u'apply_wide': 'False', u'keyword': 'x', u'title': 'X-Factor', u'value': '0', u'choices': '', u'allow_empty': 'False', u'report': 'False', u'hidden': 'False', u'result_type': 'numeric', u'unit': ''}
(Pdb++)
```

Can you take care of this?

toropok commented:
Yes! Definitely, that's what I wrote about a few posts earlier. ([here](https://github.com/senaite/senaite.core/pull/2600#issuecomment-2955123934))

> When calling calc.getInterimFields() from the debug console or in a Calculation view context, it returns the correct object. However, when the getters are invoked within the Analysis context, things are get messy.
> 
> Calculation.getInterimFields() (or Plone debug console):
> ```
> [{'apply_wide': False, 'keyword': u'TC5', 'title': u'TC5', 'value': '', 'choices': '', 'allow_empty': False, 'report': False, 'hidden': False, 'result_type': 'numeric', 'unit': ''}]
> ```
> Analysis.getInterimFields() view it becomes:
> ```
> [{u'apply_wide': u'False', u'keyword': u'TC5', u'title': u'TC5', u'value': u'', u'choices': u'', u'allow_empty': u'False', u'report': u'False', u'hidden': u'False', u'result_type': u'numeric', u'unit': u''}]
> ```

We were unable to determine the root cause of the unexpected behaviour. I mean, it looks like something wraps and slightly modifies the response when we call getInterimFields() from the AT content-type Analysis. Have you seen anything like that before?

ramonski commented:
Ah yes, I see. The calculation is now a version wrapper object and returns the JSONified values for the interims:

```python
(Pdb++) calculation
<VersionWrapper:Calculation(calculation-12@v2)>
(Pdb++) calculation.getInterimFields()
[{u'apply_wide': u'False', u'keyword': u'x', u'title': u'X-Factor', u'value': u'0', u'choices': u'', u'allow_empty': u'False', u'report': u'False', u'hidden': u'False', u'result_type': u'numeric', u'unit': u''}]
(Pdb++) calculation.content.getInterimFields()
[{'apply_wide': False, 'keyword': u'x', 'title': u'X-Factor', 'value': u'0', 'choices': '', 'allow_empty': False, 'report': False, 'hidden': False, 'result_type': 'numeric', 'unit': ''}]
```

Phew, let's see how we can solve this...

ramonski committed:
Handle version wrapped list/dict values

ramonski committed:
Merge branch 'dx-calculations' of github.com:toropok/senaite.core into dx-calculations

ramonski commented:
Ok, we handle now list/dicts in the version wrapper with this change: https://github.com/senaite/senaite.core/pull/2600/commits/2a6f07f32039ece7287282fa23e3d0433583b161

This whole thing gets more complex than expected if we do not want to store the whole object for each version...

You can give it another try now.

toropok commented:
Cool, it works! In the future, it probably makes sense to consider usage of a Schema as a reference and an extendable list of `to_field` converters, rather than listing all possible options within the `VersionWrapper.to_field_value()` method.

ramonski committed:
Merge branch '2.x' into dx-calculations

ramonski committed:
Rebuild production JS

ramonski committed:
Removed unused import

ramonski committed:
Updated metadata version

ramonski committed:
production js

ramonski committed:
Merge branch '2.x' into dx-calculations

ramonski committed:
whitespace only

ramonski commented:
Unfortunately, we have an issue here.

When changing to the `dx-calculations` branch, existing analyses no longer link to the calculation because of this change:

https://github.com/senaite/senaite.core/pull/2600/files#diff-85f0e866b6d6166de01fe6d6f109fe28ed787c8b4e51980a98efb2c256df70a3L133-R132

Therefore, these existing analyses are not correctly migrated to link to the new DX calculation object.

I guess we need to run a migration over all existing analyses to migrate the reference fields.
Or maybe we can use the `portal_references` catalog for that 🤔

ramonski committed:
Migrate Analyses Calculations

ramonski committed:
Delete Analysis Calculation Reference

ramonski committed:
Added comment

ramonski committed:
Compatibility AT/DX snapshots

ramonski committed:
Compatibility with AT based snapshots

ramonski committed:
Test fixture

ramonski committed:
typo

ramonski committed:
Another test fixture

ramonski committed:
fixed duplicate variable name

ramonski committed:
Another test fixture

ramonski committed:
Removed test module

Commit:
make deep_get simplier and faster

xispa commented:
> Or maybe we can use the `portal_references` catalog for that 🤔

This would be more performant and would also give us the opportunity to clean-up the reference catalog of this calculation-analysis references at the same time

ramonski committed:
Merge branch '2.x' into dx-calculations

ramonski commented:
> > Or maybe we can use the `portal_references` catalog for that 🤔
> 
> This would be more performant and would also give us the opportunity to clean-up the reference catalog of this calculation-analysis references at the same time

I did that already here:
https://github.com/senaite/senaite.core/blob/aa131b07ae045eb04cbfb2e0a596f37ee9d04d41/src/senaite/core/upgrade/v02_07_000.py#L286-L298

So besides one orphan entry from a broken sample object, the catalog was clean after the migration...

ramonski committed:
Merge branch '2.x' into dx-calculations

ramonski committed:
Merge branch 'dx-calculations' of github.com:toropok/senaite.core into dx-calculations

ramonski committed:
Merge branch '2.x' into dx-calculations

ramonski committed:
Updated

ramonski committed:
Updated Production JS

ramonski committed:
Bumped upgrade step version

ramonski committed:
removed versionwrapper

ramonski committed:
Upgrade step updated

ramonski committed:
removed versionwrapper adapter

ramonski committed:
removed referencefield version awareness

ramonski committed:
Added fields to store the formula and imports

ramonski committed:
Handle calculation set

ramonski committed:
use field to access value

ramonski committed:
updated test

ramonski committed:
Merge branch '2.x' into dx-calculations

ramonski committed:
DependentServices -> dependent_services

ramonski committed:
Handle nonexisting fields

ramonski committed:
Convert None Values and normalize type

ramonski committed:
Call setCalculation

ramonski committed:
Service popup fixed

ramonski committed:
Moved methods to abstractanalysis

ramonski committed:
Better error message

ramonski committed:
Added new exported demo data package

ramonski committed:
fixed interims type -> result_type in tests

ramonski committed:
Check for routine analysis

ramonski committed:
move back to abstractanalysis

ramonski committed:
Reindex after set

ramonski committed:
Fix tests

ramonski committed:
Renaming to getCalculationDependents

ramonski committed:
Fixed types -> result_types interims

ramonski committed:
Test getCalculationFormula instead

ramonski committed:
Fixed after_retest

ramonski committed:
Removed unused imports

ramonski committed:
removed redefinition

ramonski committed:
Removed unused import

ramonski committed:
Merge branch '2.x' into dx-calculations

ramonski committed:
Webpack build

ramonski committed:
Fix interim validator

Only assume that same interim keywords must have same titles, not vice
versa. Also only consider active calculations.

ramonski committed:
Use API to get the UID

ramonski committed:
remove calc interims from services

ramonski committed:
test fixture

ramonski committed:
Calc interims first

ramonski committed:
Don't set service interims

ramonski committed:
Always use interims from source

ramonski committed:
Merge calc interims with service interims

ramonski committed:
Return interims as they were set

ramonski committed:
handle apply_wide vs wide

ramonski committed:
docstring fixed

ramonski committed:
removed historyawarereferencefield

ramonski committed:
removed dependency to CMFEditions

ramonski committed:
Removed unused function

ramonski committed:
Removed unused parameter

ramonski committed:
migrate apply_wide

ramonski committed:
Added import validator

ramonski committed:
Fixed migration of interim flags

ramonski committed:
fixed missing parameter

xispa reviewed (approved):
This is extreme work!, thanks @ramonski @toropok . Let's settle it down a bit.
